test/test_torch.py

```diff
@@ -4341,6 +4341,90 @@ def test_diff(self, device, dtype):
+        inputs = (
+            ((4, 4, 4), [(1., 3.0, 6., 10.), (1., 3.0, 4., 90.), (1., 2.5, 3.4, 7.)], [0, 1, 2], 1),
+            ((4, 4, 4, 3), [(1., 3.0, 6., 10.), (1., 2.5, 7.)], [0, -1], 1),
+            ((4, 5, 2, 4, 3), [(1., 3.0, 6., 10.), (1., 3.4, 7.)], [-2, -1], 1),
+            ((4, 3, 6, 5, 3), [(1., 3.0, 6., 10.), (2.5, 3.4, 7.)], [0, 1], 1),
+            ((4, 3, 2, 5, 3), [(1., 3.0), (3.4, 7., 6.0)], [2, 1], 1),
+            ((4, 4, 4, 3), [1.23, (1., 2.5, 3.4,)], [0, -1], 1),
+            ((4, 5, 2, 4, 3), [(1., 3.0, 6., 10.), 3.45], [-2, -1], 1),
+        )
+
+        for size, spacing, axis, edge_order in inputs:
+            for contig in [True, False]:
+                t = make_tensor(size, device, dtype, noncontiguous=not contig)
+                t_n = t.cpu().numpy()
+                spacing_list = []
+                for space in spacing:
+                    spacing_list.append(torch.tensor(space, dtype=float, device=device))
+                result = torch.gradient(t, spacing=spacing_list, dim=axis, edge_order=edge_order)
+                result_n = np.gradient(t_n, *spacing, axis=axis, edge_order=edge_order)
+                if not isinstance(result_n, list):
+                    result_n = [result_n]
+                self.assertEqual(result, result_n)
+
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.long, torch.float32, torch.complex64)
+    def test_gradient_scalar(self, device, dtype):
+        inputs = (
+            ((5,), [2.73], 0, 1),
+            ((5, 5), 2.0, 0, 1),
+            ((2, 2, 2), [1.17, 3.34, 1.73], [0, 1, 2], 1),
+            ((4, 4, 4), 1.7 , [0, 1, 2], 1),
+            ((4, 3, 4), 1.7 , [-2, -1], 1),
+            ((4, 3, 4, 5), [1.7, 3.93], [0, -1], 1),
+            ((4, 3, 4, 5), 1.7 , [1, 2], 1),
+            ((4, 3, 4, 5), 1.7 , [0, 1], 1),
+            ((4, 3, 4, 5), 1.7 , [2, -1], 1),
+        )
+
+        for size, spacing, axis, edge_order in inputs:
+            for contig in [True, False]:
+                t = make_tensor(size, device, dtype, noncontiguous=not contig)
+                t_n = t.cpu().numpy()
+                result = torch.gradient(t, spacing=spacing, dim=axis, edge_order=edge_order)
+                spacing_ = spacing if isinstance(spacing, list) else [spacing]
+                result_n = np.gradient(t_n, *spacing_, axis=axis, edge_order=edge_order)
+                if not isinstance(result_n, list):
+                    result_n = [result_n]
+                self.assertEqual(result, result_n)
+
+        result = torch.gradient(torch.tensor([2, -2, inf, inf, -inf, -inf, inf, 3, -inf, 2, nan, nan, 3, inf, nan]))
+        result_n = np.gradient(np.array([2, -2, inf, inf, -inf, -inf, inf, 3, -inf, 2, nan, nan, 3, inf, nan]))
+        if not isinstance(result_n, list):
+            result_n = [result_n]
+        self.assertEqual(result, result_n,)
+
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.long, torch.float32, torch.complex64)
```

Comment:
These tests are pretty good; can we also add a test with some large tensors and randomly generated spacing, and a test with large tensors for when spacing values are negative, zero, and positive.

By "large" a few thousand elements should do it.